File: py/btree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class BNode:
	def __init__(self, order):
		self.order = order
		self.order_2 = order*2
		self.order_3 = order*2+1
		self.recs = []
		self.childs = []

	def __str__(self):
		output = []

		for i in xrange(self.order_3):
			if i < len(self.childs): output.append(str(self.childs[i]))
			if i < len(self.recs): output.append('[%s]' % str(self.recs[i]))

		return '(%s)' % ' '.join(output)

	def insert_rec(self, rec, parents=[], parent_is=[]):
		cur_i = len(self.recs)
		for i, _rec in enumerate(self.recs):
			if _rec > rec:
				cur_i = i
				break
			elif _rec == rec:
				raise ValueError

		if self.childs:
			self.childs[cur_i].insert_rec(rec, parents+[self], parent_is+[cur_i])
		else:
			self.recs.insert(cur_i, rec)
			self.insert_rec_adjust(parents, parent_is)

	def insert_rec_adjust(self, parents, parent_is):
		parent = parents[-1] if parents else None
		parent_i = parent_is[-1] if parent_is else -1

		if len(self.recs) <= self.order_2:
			return

		mid_rec = self.recs[len(self.recs)/2]

		left = BNode(self.order)
		left.recs = self.recs[:len(self.recs)/2]
		left.childs = self.childs[:len(self.childs)/2]

		right = BNode(self.order)
		right.recs = self.recs[len(self.recs)/2+1:]
		right.childs = self.childs[len(self.childs)/2:]

		self.childs = [left, right]

		if parent:
			parent.recs = parent.recs[:parent_i] + [mid_rec] + parent.recs[parent_i:]
			parent.childs = parent.childs[:parent_i] + [left, right] + parent.childs[parent_i+1:]
			parent.insert_rec_adjust(parents[:-1], parent_is[:-1])
		else:
			self.recs = [mid_rec]

	def remove_rec(self, rec, parents=[], parent_is=[]):
		for i, _rec in enumerate(self.recs):
			if _rec > rec:
				if self.childs: return self.childs[i].remove_rec(rec, parents+[self], parent_is+[i])
				else: raise ValueError
			elif _rec == rec:
				if not self.childs:
					self.recs[i:i+1] = []
					return self.remove_rec_adjust(parents, parent_is)
				else:
					prev_i = i+1
					prev_suc = self
					suc = prev_suc.childs[prev_i]

					new_parents = parents + [prev_suc]
					new_parent_is = parent_is + [prev_i]

					while suc.childs:
						prev_suc = suc
						prev_i = 0
						suc = suc.childs[prev_i]

						new_parents.append(prev_suc)
						new_parent_is.append(prev_i)

					_rec = suc.recs[0]

					self.recs[i] = _rec
					return suc.remove_rec(_rec, new_parents, new_parent_is)
		else:
			if self.childs: return self.childs[-1].remove_rec(rec, parents+[self], parent_is+[len(self.childs)-1])
			else: raise ValueError

	def remove_rec_adjust(self, parents, parent_is):
		parent = parents[-1] if parents else None
		parent_i = parent_is[-1] if parent_is else -1

		if not parent:
			if not self.recs and self.childs: return self.childs[0]
			return

		elif len(self.recs) >= self.order:
			return

		elif parent_i+1 < len(parent.childs) and len(parent.childs[parent_i+1].recs) > self.order:
			self.recs.append(parent.recs[parent_i])
			if parent.childs[parent_i+1].childs:
				self.childs.append(parent.childs[parent_i+1].childs[0])
				parent.childs[parent_i+1].childs[0:1] = []

			parent.recs[parent_i] = parent.childs[parent_i+1].recs[0]
			parent.childs[parent_i+1].recs[0:1] = []

		elif parent_i-1 >= 0 and len(parent.childs[parent_i-1].recs) > self.order:
			self.recs.insert(0, parent.recs[parent_i-1])
			if parent.childs[parent_i-1].childs:
				self.childs.insert(0, parent.childs[parent_i-1].childs[-1])
				parent.childs[parent_i-1].childs[-1:] = []

			parent.recs[parent_i-1] = parent.childs[parent_i-1].recs[-1]
			parent.childs[parent_i-1].recs[-1:] = []

		elif parent_i+1 < len(parent.childs):
			self.recs = self.recs + [parent.recs[parent_i]] + parent.childs[parent_i+1].recs
			self.childs += parent.childs[parent_i+1].childs

			parent.recs[parent_i:parent_i+1] = []
			parent.childs[parent_i+1:parent_i+2] = []

			return parent.remove_rec_adjust(parents[:-1], parent_is[:-1])

		elif parent_i-1 >= 0:
			self.recs = parent.childs[parent_i-1].recs + [parent.recs[parent_i-1]] + self.recs
			self.childs = parent.childs[parent_i-1].childs + self.childs

			parent.recs[parent_i-1:parent_i] = []
			parent.childs[parent_i-1:parent_i] = []

			return parent.remove_rec_adjust(parents[:-1], parent_is[:-1])

		else:
			assert False

	def search_rec(self, key):
		for i, _rec in enumerate(self.recs):
			if _rec > key:
				if self.childs: return self.childs[i].search_rec(key)
				else: raise ValueError
			elif _rec == key:
				return _rec
		else:
			if self.childs: return self.childs[-1].search_rec(key)
			else: raise ValueError

class BTree:
	def __init__(self, order):
		self.order = order
		self.root = BNode(self.order)

	def __str__(self):
		return str(self.root)

	def insert_rec(self, rec):
		self.root.insert_rec(rec)

	def remove_rec(self, rec):
		new_root = self.root.remove_rec(rec)
		if new_root: self.root = new_root

	def search_rec(self, rec):
		return self.root.search_rec(rec)

File: hs/BTree.hs
module BTree (BTree(BNode), tree, recs, childs, insert, remove, search) where

import qualified Data.List as List

data BTree rec = BNode Int [rec] [BTree rec]

_show [] [] = []
_show [] childs =
	show (head childs) :
	_show [] (tail childs)
_show recs [] =
	("[" ++ show (head recs) ++ "]") :
	_show (tail recs) []
_show recs childs =
	show (head childs) :
	("[" ++ show (head recs) ++ "]") :
	_show (tail recs) (tail childs)

instance (Show a) => Show (BTree a) where
	show (BNode _ recs childs) = "(" ++ List.intercalate " " (_show recs childs) ++ ")"

tree order = BNode order [] []
recs (BNode _ recs _) = recs
childs (BNode _ _ childs) = childs

insert rec (BNode order recs childs) =
	if not (null childs)
		then
			let BNode _ child_recs child_childs = insert rec $! (childs !! cur_i)
			in
				if length child_recs == 1 && length (BTree.recs (childs !! cur_i)) /= 1
					then insert_adjust $! (BNode order
						((take cur_i recs) ++ [head child_recs] ++ (drop cur_i recs))
						((take cur_i childs) ++ child_childs ++ (drop (cur_i+1) childs)))
					else BNode order
						recs
						((take cur_i childs) ++ [BNode order child_recs child_childs] ++ (drop (cur_i+1) childs))
		else
			insert_adjust $! (BNode order
				((take cur_i recs) ++ [rec] ++ (drop cur_i recs))
				childs)
	where cur_i = case List.findIndex (>= rec) recs of
		Just i -> i
		Nothing -> length recs

insert_adjust (BNode order recs childs) =
	if length recs > order*2
		then BNode order
			[recs !! order]
			[BNode order
				(take order recs)
				(take (length childs `div` 2) childs),
			BNode order
				(drop (order+1) recs)
				(drop (length childs `div` 2) childs)]
		else BNode order recs childs

remove rec (BNode order recs childs)
	| not (null (BTree.recs node)) = node
	| otherwise =
		if not (null (BTree.childs node))
			then head (BTree.childs node)
			else node
	where node = remove2 rec (BNode order recs childs)

remove2 rec (BNode order recs childs) =
	if not (null childs)
		then
			if cur_i < length recs && recs !! cur_i == rec
				then
					let suc = remove_suc (childs !! (cur_i+1))
					in remove_adjust (cur_i+1) (BNode order
						((take cur_i recs) ++ [suc] ++ (drop (cur_i+1) recs))
						((take (cur_i+1) childs) ++ [(remove2 suc (childs !! (cur_i+1)))] ++ (drop (cur_i+2) childs)))
				else remove_adjust cur_i (BNode order
					recs
					((take cur_i childs) ++ [(remove2 rec (childs !! cur_i))] ++ (drop (cur_i+1) childs)))
		else BNode order
			((take cur_i recs) ++ (drop (cur_i+1) recs))
			childs
	where cur_i = case List.findIndex (>= rec) recs of
		Just i -> i
		Nothing -> length recs

remove_adjust cur_i (BNode order recs childs) =
	if length (BTree.recs child) < order
		then
			if cur_i+1 < length childs && length (BTree.recs right) > order then
				let
					rec = recs !! cur_i
					right_rec = head (BTree.recs right)
				in BNode order
					((take cur_i recs) ++ [right_rec] ++ (drop (cur_i+1) recs))
					((take cur_i childs) ++
						[BNode order
							(BTree.recs child ++ [rec])
							(BTree.childs child ++ if not (null (BTree.childs right)) then [head (BTree.childs right)] else []),
						BNode order
							(tail (BTree.recs right))
							(if not (null (BTree.childs right)) then tail (BTree.childs right) else [])] ++
						(drop (cur_i+2) childs))
			else if cur_i-1 >= 0 && length (BTree.recs left) > order then
				let
					rec = recs !! (cur_i-1)
					left_rec = last (BTree.recs left)
				in BNode order
					((take (cur_i-1) recs) ++ [left_rec] ++ (drop cur_i recs))
					((take (cur_i-1) childs) ++
						[BNode order
							(init (BTree.recs left))
							(if not (null (BTree.childs left)) then init (BTree.childs left) else []),
						BNode order
							(rec : BTree.recs child)
							((if not (null (BTree.childs left)) then [last (BTree.childs left)] else []) ++ BTree.childs child)] ++
						(drop (cur_i+1) childs))
			else if cur_i+1 < length childs then
				let rec = recs !! cur_i
				in BNode order
					((take cur_i recs) ++ (drop (cur_i+1) recs))
					((take cur_i childs) ++
						[BNode order
							(BTree.recs child ++ [rec] ++ BTree.recs right)
							(BTree.childs child ++ BTree.childs right)] ++
						(drop (cur_i+2) childs))
			else if cur_i-1 >= 0 then
				let rec = recs !! (cur_i-1)
				in BNode order
					((take (cur_i-1) recs) ++ (drop cur_i recs))
					((take (cur_i-1) childs) ++
						[BNode order
							(BTree.recs left ++ [rec] ++ BTree.recs child)
							(BTree.childs left ++ BTree.childs child)] ++
						(drop (cur_i+1) childs))
			else
				BNode order [] [] -- assert False
		else BNode order recs childs
	where
		child = childs !! cur_i
		right = childs !! (cur_i+1)
		left = childs !! (cur_i-1)

remove_suc (BNode order recs childs)
	| not (null childs) = remove_suc (head childs)
	| otherwise = head recs

search rec (BNode order recs childs) =
	case List.findIndex (>= rec) recs of
		Just i ->
			if recs !! i == rec
				then Just (recs !! i)
				else
					if i < length childs
						then search rec (childs !! i)
						else Nothing
		Nothing ->
			if not (null childs)
				then search rec (last childs)
				else Nothing
